### DISCUSSION TITLE
cannot remove/replace a hydrogem with stereochem info

### DISCUSSION THREAD

**Discussion by YunchaoLiu** [rdkit/rdkit#4653] [Q&A]
Hello, I am trying to replace a hydrogen with a carbon. But I tried to delete it, or replace it but cannot get either of them to work. Here is the smiles: Cc1onc(-c2ccccc2)c1C(=O)N[C@@H]1C(=O)N2[C@@H](C(=O)O)C(C)(C)S[C@H]12, atom #27 is stereo center and connected to a H.

When deleting, I tried

1.
` mol = Chem.RemoveAllHs(mol)`

2. 
```
params = Chem.rdmolops.RemoveHsParameters()
params.removeWithWedgedBond=True
params.removeDefiningBondStereo = True
mol = Chem.RemoveHs(mol, params)
```

3. 
```
mol = Chem.rdmolops.RemoveStereochemistry(mol) 
mol.UpdatePropertyCache() # this one still shows  explicit_Hs = 1
```

I also try to replace it. However, I cannot even find the H by using 
1. 
`atom.GetNeighors()`

2.
```
for bond in atom.GetBonds():
   if (bond.GetBeginAtom().GetAtomicNum() ==1) or (bond.GetEndAtom().GetAtomicNum() ==1):
                     print('replaced') # this never gets printed. I checked and it's because the bond connected to the H is not even shown in atom.GetBonds().
```

**Comment by dehaenw** [comment]
Hello @YunchaoLiu 
Could it be that there is something wrong with your SMILES string? If i run the following with a corrected SMILES, I get the expected output (no explicit Hs to begin with)
```python
import rdkit
from rdkit import Chem
mol = Chem.MolFromSmiles("Cc1onc(-c2ccccc2)c1C(=O)N[C@@H]1C(=O)N2[C@@H]C(C)(C)S[C@H]12")
for atom in mol.GetAtoms():
    if atom.GetAtomicNum()==1:
        print("there is a hydrogen")
```
whereas if i run 
```python
mol = Chem.MolFromSmiles("Cc1onc(-c2ccccc2)c1C(=O)N[C@@h]1C(=O)N2C@@HC(C)(C)S[C@H]12")
```
i get a parsing error and a None mol returned

**Reply by YunchaoLiu** [reply]
Hi, @dehaenw , thanks for the answer. As @ptosco pointed out below, I found the corrupted smiles was due to the my wrong formatting in github. His answer solves the problem. Thank you anyway :)

**Comment by ptosco** [comment]
@YunchaoLiu Make sure you always write SMILES strings inside backticks or they will come up corrupted in GitHub.
Your molecule has no hydrogen atoms in the molecule graph:
```
from rdkit import Chem

mol = Chem.MolFromSmiles("Cc1onc(-c2ccccc2)c1C(=O)N[C@@H]1C(=O)N2[C@@H]C(C)(C)S[C@H]12")
for a in mol.GetAtoms():
    print(f"{a.GetIdx():2d} {a.GetSymbol():>2s} {a.GetNumImplicitHs():2d} {a.GetNumExplicitHs():2d}")
 0  C  3  0
 1  C  0  0
 2  O  0  0
 3  N  0  0
 4  C  0  0
 5  C  0  0
 6  C  1  0
 7  C  1  0
 8  C  1  0
 9  C  1  0
10  C  1  0
11  C  0  0
12  C  0  0
13  O  0  0
14  N  1  0
15  C  0  1
16  C  0  0
17  O  0  0
18  N  0  0
19  C  1  0
20  C  0  0
21  C  3  0
22  C  3  0
23  S  0  0
24  C  0  1
```
It only has implicit and explicit Hs, i.e. virtual hydrogen atoms which are not actually included in the molecule graph as real atoms connected by real bonds, but only as counts associated to their parent heavy atom.
The "implicit" and "explicit" terminology in RDKit has a different meaning from the one normally used in cheminformatics; in fact, both implicit and explicit Hs are in a wider sense implicit. This is why they do not show up in the atom list nor in the bond list.
Implicit Hs are those automatically added by RDKit to heavy atoms to fill the valences not occupied by heavy atoms.
Instead, when you specify Hs for a heavy atom inside of square brackets in the SMILES string, e.g. to specify a tautomer or as in your case to specify stereochemistry, they become explicit Hs and are counted as such - but they are still not real hydrogens.

If you wish to add real hydrogens to the molecule graph, use [`Chem.AddHs()`](http://rdkit.org/docs/source/rdkit.Chem.rdmolops.html?highlight=addhs#rdkit.Chem.rdmolops.AddHs). If you only want to convert explicit Hs into real hydrogens, set `explicitOnly` to `True`. If you want to convert both implicit Hs and explicit Hs into real hydrogens, set `explicitOnly` to `False` (the default).

The topic of implicit and explicit Hs in RDKit has already come up many times on the RDKit mailing list; you may read more here:
https://sourceforge.net/p/rdkit/mailman/message/29679834/
https://sourceforge.net/p/rdkit/mailman/message/36696340/

There is also a very clear and well-written blog post by Roger Sayle:
https://nextmovesoftware.com/blog/2013/02/27/explicit-and-implicit-hydrogens-taking-liberties-with-valence/

**Reply by YunchaoLiu** [reply]
Thank you @ptosco ! It worked
